### ISSUE TITLE
ActiveStarMap needs unit tests

### ISSUE THREAD

**Issue by borg286** [HellFirePvP/AstralSorcery#761]
Looking over the logic for ActiveStarMap I see a whole bunch of heuristics for calculating the scaling factor for enchantments based on the intersection of the constellation bounding boxes. But the logic is hard to follow and seems error prone. You should add some unit tests for this class so that some toy examples of input bounding boxes produce the scaling factor you want.
Since the randomness is only used for calculating the duration on potions it should be deterministic.
I recommend you walk through the code using 3 perfectly overlapping constellations and see if shiftDistribution does what you want. Based on my calculations this line of code
https://github.com/HellFirePvP/AstralSorcery/blob/8611d9558db776cf74d96b0b9be3b1377a471ea0/src/main/java/hellfirepvp/astralsorcery/common/constellation/starmap/ActiveStarMap.java#L90
will result in perfectly overlapping constellations to have a final perc value of 3, which gets clamped to 1, which gets shifted to .44.

The problem I suspect is that because you perform this loop
https://github.com/HellFirePvP/AstralSorcery/blob/8611d9558db776cf74d96b0b9be3b1377a471ea0/src/main/java/hellfirepvp/astralsorcery/common/constellation/starmap/ActiveStarMap.java#L74
a total of 3 times that intersectingSize gets unwanted values.  

I recommend you simply take the first constellation and intersect it with the 2nd constellation (using a default rectangle of the entire canvas. Then take that intersection and intersect it with the 3rd constellation, call this the final intersection. Calculate the bounding box for all 3 constelations. Take that final intersection area and divide by the bounding box area and call that the final perc. This will make it much easier to reason. If you want go ahead and shift it, but I don't think that is necessary.

**Comment by borg286** [comment]
I also propose an alternative for calculating the percent value. Instead of having a fixed scaling factor based on the placement of the constellations, instead the constellations will simply be selected, and the drawing board simply shows how bright the contellations are in the sky. The current block favors gaming the system, and gives the false impression that the user should try to align the start. The enchantment level seems to have nothing to do with overlapping start, but only overlapping rectangles. Once the player knows that this is the only thing that matters they will simply stack them, and thus removes the fun.
The question is how to retain the 'drawing' aspect, while letting players chose how to balance between the enchantments from one constellation and another. The current UI does not give the feedback as to how aligned the sky is with the selected constellations.
I propose that the constellation that is at the top of the page to be the one who's enchantments are weighted the most, and drop rectangle overlap altogether. Have the scalaing factor be based purely on how much that constellation rules the sky.

**Comment by yagomont** [comment]
Hey, thanks for bringing this up. Reading the astral tome, i'd have never figured out how placing the constellations on the parchment would influence the resulting effects from the engraved glass. But, i'm still quite confused. Right now as the code is, are you supposed to stack constellations on each other, stars, or leave them apart as much as you can?

**Comment by Doomgull** [comment]
@yagomont spread them apart, unless using a version more than a year old.

Also, Page 2 of the Stellar Refraction Table in the Astral Tome:
"The placement of the Constellations seems to have great a impact as to their dominance in the final pattern. The fewer overlapping points from each Constellation, the stronger their presence will be in the final drawing."

**Comment by yagomont** [comment]
I see. What was unclear to me was if by points it meant the stars or the constellation lines, or both. Apparently keeping them not touching is enough.

**Comment by yagomont** [comment]
Uhh... What the fuck.

I just made 8 engraved glass and tried different combinations to test this, with varying levels of distance between the stars. What I saw was that stacking constellations on each other actually produced better results than if they are not touching at all!

Armara not touching its neighbours, for example, gave me prot IV consistently.

Now, stacked upon Pelotrio and Mineralis, it gave prot V.

Repeat the same happenings in a few other tries.

Thus, experimentation brings me to this conclusion:

This is broken.
